File: src/main/java/foundry/animamundi/AnimaMundi.java
package foundry.animamundi;

import com.mojang.logging.LogUtils;
import com.tterrag.registrate.Registrate;
import com.tterrag.registrate.util.nullness.NonNullSupplier;
import foundry.animamundi.content.compat.alembic.CodecUtil;
import foundry.animamundi.grimoire.AnimaMundiEntities;
import foundry.animamundi.grimoire.AnimaMundiItems;
import foundry.animamundi.grimoire.AnimaMundiNetworking;
import net.minecraft.client.Minecraft;
import net.minecraft.client.player.LocalPlayer;
import net.minecraft.core.BlockPos;
import net.minecraft.core.Vec3i;
import net.minecraft.network.chat.Component;
import net.minecraft.resources.ResourceLocation;
import net.minecraft.util.Mth;
import net.minecraft.world.level.ClipContext;
import net.minecraft.world.level.Level;
import net.minecraft.world.level.block.Blocks;
import net.minecraft.world.level.block.state.BlockState;
import net.minecraft.world.level.levelgen.structure.BoundingBox;
import net.minecraft.world.phys.BlockHitResult;
import net.minecraft.world.phys.Vec3;
import net.minecraftforge.common.MinecraftForge;
import net.minecraftforge.event.TickEvent;
import net.minecraftforge.eventbus.api.IEventBus;
import net.minecraftforge.eventbus.api.SubscribeEvent;
import net.minecraftforge.fml.common.Mod;
import net.minecraftforge.fml.event.lifecycle.FMLClientSetupEvent;
import net.minecraftforge.fml.event.lifecycle.FMLCommonSetupEvent;
import net.minecraftforge.event.server.ServerStartingEvent;
import net.minecraftforge.fml.javafmlmod.FMLJavaModLoadingContext;
import org.slf4j.Logger;

import javax.swing.*;
import java.util.HashMap;
import java.util.Map;

import static org.openjdk.nashorn.internal.objects.ArrayBufferView.length;

@Mod(AnimaMundi.MODID)
public class AnimaMundi {

    public static final String MODID = "animamundi";
    private static final Logger LOGGER = LogUtils.getLogger();
    public static final NonNullSupplier<Registrate> REGISTRATE = NonNullSupplier.lazy(() -> Registrate.create(MODID));

    public static Registrate registrate() {
        return REGISTRATE.get();
    }

    public static ResourceLocation path(String path) {
        return new ResourceLocation(MODID, path);
    }

    public AnimaMundi() {
        MinecraftForge.EVENT_BUS.register(this);
        IEventBus modEventBus = FMLJavaModLoadingContext.get().getModEventBus();
        CodecUtil.init();
        AnimaMundiItems.INSTANCE.register();
        AnimaMundiEntities.register();
        AnimaMundiNetworking.init();
    }

    private void commonSetup(final FMLCommonSetupEvent event) {

    }

    @SubscribeEvent
    public void onServerStarting(ServerStartingEvent event) {

    }

    @SubscribeEvent
    public void onServerTick(TickEvent.ServerTickEvent event) {

    }

    @Mod.EventBusSubscriber(modid = MODID, bus = Mod.EventBusSubscriber.Bus.MOD)
    public static class ClientModEvents {

        @SubscribeEvent
        public static void onClientSetup(FMLClientSetupEvent event) {

        }
    }
}

File: src/main/java/foundry/animamundi/AnimaMundi.kt
package foundry.animamundi

import com.mojang.logging.LogUtils
import com.tterrag.registrate.Registrate
import com.tterrag.registrate.util.nullness.NonNullSupplier
import foundry.animamundi.AnimaMundi
import foundry.animamundi.content.compat.alembic.CodecUtil
import foundry.animamundi.grimoire.AnimaMundiEntities
import foundry.animamundi.grimoire.AnimaMundiItems.register
import foundry.animamundi.grimoire.AnimaMundiNetworking
import foundry.animamundi.happenings.ForgeCommonEvents
import net.minecraft.resources.ResourceLocation
import net.minecraftforge.common.MinecraftForge
import net.minecraftforge.event.TickEvent.ServerTickEvent
import net.minecraftforge.event.server.ServerStartingEvent
import net.minecraftforge.eventbus.api.SubscribeEvent
import net.minecraftforge.fml.common.Mod
import net.minecraftforge.fml.common.Mod.EventBusSubscriber
import net.minecraftforge.fml.event.lifecycle.FMLClientSetupEvent
import net.minecraftforge.fml.event.lifecycle.FMLCommonSetupEvent
import net.minecraftforge.fml.javafmlmod.FMLJavaModLoadingContext
import thedarkcolour.kotlinforforge.forge.MOD_BUS
import thedarkcolour.kotlinforforge.forge.runForDist

@Mod(AnimaMundi.MODID)
object AnimaMundi {
    public const val MODID = "animamundi"
    public val LOGGER = LogUtils.getLogger()
    public val REGISTRATE = NonNullSupplier.lazy { Registrate.create(MODID) }

    fun registrate(): Registrate {
        return REGISTRATE.get()
    }

    fun path(path: String?): ResourceLocation {
        return ResourceLocation(MODID, path)
    }

    init {
        CodecUtil.init()
        register()
        AnimaMundiEntities.register()
        AnimaMundiNetworking.init()

        val obj = runForDist(
            clientTarget = {},
            serverTarget = {
                MOD_BUS.addListener(ForgeCommonEvents::attachCaps)
                MOD_BUS.addListener(ForgeCommonEvents::playerTick)
            }
        )
    }

}